specs/quota-custom-resource-class.rst

```diff
@@ -0,0 +1,246 @@
+..
+ This work is licensed under a Creative Commons Attribution 3.0 Unported
+ License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+
+===============================
+Quota per custom resource class
+===============================
+
+https://blueprints.launchpad.net/nova/+spec/tbd
+
+With this blueprint we introduce a way to define and use quotas per custom
+resources classes, so that operator can limit how DC resources are utilized.
+
+Problem description
+===================
+
+Currently, quotas in Nova are restricted only for selected resources like CPU
+cores, ram, floating IPs, and for non-resource things: number of instances,
+metadata items, key pairs, server groups and so on. Introducing `custom
+resource classes`_ within placement service, allows operator to create
+resources, which are unknown to the quota mechanism.
+
+Use Cases
+---------
+
+As an operator I’d like to define quota for custom resource classes, so that
+amount of such resources can be limited per project and/or per user.
+
+Proposed change
+===============
+
+We propose to add ability for dynamically adding custom resource classes to
+resources which ``QuotaEngine`` can handle, and track the usage like all the
+other resources.
```

Comment:
So the quota limit checking is done by access the Placement API? Who is going to report the inventory for those custom resource class?

Reply:
Information of inventory and allocations can be obtained from Placement API. My guess is, that we should check allocations vs the quotas against certain classes during request handling in nova API.